### PR TITLE
support CastTimeAsDuration in TiFlash

### LINKED ISSUE #5306: support `CastTimeAsDuration` in TiFlash
## Feature Request

**Is your feature request related to a problem? Please describe:**

**Describe the feature you'd like:**

Support `CastTimeAsDuration` in TiFlash
TiDB's implementation:
https://github.com/pingcap/tidb/blob/562bb9f9e2c194c3f398300d5e5e920965da9f47/expression/builtin_cast.go#L1454
**Describe alternatives you've considered:**

**Teachability, Documentation, Adoption, Migration Strategy:**

### PR DESCRIPTION
**Pull request by AntiTopQuark**
### What problem does this PR solve?

Issue Number: close #5306
related pr:https://github.com/pingcap/tidb/pull/37036

Problem Summary:
support CastTimeAsDuration in TiFlash

### What is changed and how it works?
 1. register the cast function
 2. implement the calculation for casting time to duration
 3. add tests of conversion
### Check List

Tests 

- [x] Unit test
- [x] Integration test
- [ ] Manual test (add detailed scripts or steps below)
- [ ] No code

Side effects

- [ ] Performance regression: Consumes more CPU
- [ ] Performance regression: Consumes more Memory
- [ ] Breaking backward compatibility

Documentation

- [ ] Affects user behaviors
- [ ] Contains syntax changes
- [ ] Contains variable changes
- [ ] Contains experimental features
- [ ] Changes MySQL compatibility

### Release note

```release-note
Support to pushdown CastTimeAsDuration to TiFlash
```

### CODE REVIEW

File: dbms/src/Functions/FunctionsTiDBConversion.h [hunk 1]
```diff
@@ -1610,6 +1611,47 @@ struct TiDBConvertToDuration
                 block.getByPosition(result).column = std::move(to_col);
             }
         }
+        else if constexpr (std::is_same_v<FromDataType, DataTypeMyDate> || std::is_same_v<FromDataType, DataTypeMyDateTime>)
+        {
```
Comment (ywqzzy) [inline dbms/src/Functions/FunctionsTiDBConversion.h:1615]: can you split the execute function by FromDataType?

File: dbms/src/Functions/FunctionsTiDBConversion.h [hunk 2]
```diff
@@ -1610,6 +1611,47 @@ struct TiDBConvertToDuration
...
+            vec_to.resize(size);
+            const auto & to_type = checkAndGetDataType<DataTypeMyDuration>(removeNullable(block.getByPosition(result).type).get());
+            int to_fsp = to_type->getFsp();
+
+            for (size_t i = 0; i < size; ++i)
+            {
+                MyDateTime datetime(from_vec[i]);
+                if constexpr (std::is_same_v<ToDataType, DataTypeMyDate>)
+                {
+                    vec_to[i] = 0;
```
[the remark this answers is not in the document]
Comment (AntiTopQuark) [reply dbms/src/Functions/FunctionsTiDBConversion.h]: done

File: dbms/src/Functions/FunctionsTiDBConversion.h (hunk 1 above)
Comment (AntiTopQuark) [reply dbms/src/Functions/FunctionsTiDBConversion.h:1627]: done

File: dbms/src/Functions/FunctionsTiDBConversion.h [hunk 3]
```diff
@@ -1610,6 +1611,48 @@ struct TiDBConvertToDuration
...
+            }
+            block.getByPosition(result).column = std::move(to_col);
+        }
+        else if constexpr (std::is_same_v<FromDataType, DataTypeMyDateTime>)
+        {
+            // cast time as duration
+            const auto * col_from = checkAndGetColumn<ColumnUInt64>(block.getByPosition(arguments[0]).column.get());
+            const ColumnUInt64::Container & from_vec = col_from->getData();
+            const auto & col_with_type_and_name = block.getByPosition(arguments[0]);
+            const auto & from_type = static_cast<const DataTypeMyDateTime &>(*col_with_type_and_name.type);
```
Comment (SeaRise) [inline dbms/src/Functions/FunctionsTiDBConversion.h:1632]: how about using `checkAndGetDataType`?

File: tests/fullstack-test/expr/cast_as_duration.test [hunk 4]
```diff
@@ -0,0 +1,88 @@
...
+# limitations under the License.
+
+
+mysql> drop table if exists test.t;
+mysql> create table test.t(c1 date, c2 datetime(4));
+mysql> insert into test.t values('2022-01-01','2022-09-20 11:11:11.0000');
+mysql> insert into test.t values('2022-01-01','2022-09-20 11:11:11.1111');
+mysql> insert into test.t values('2022-01-01','2022-09-20 11:11:11.1234');
+mysql> insert into test.t values('2022-01-01','2022-09-20 11:11:11.1255');
+mysql> insert into test.t values('2022-01-01','2022-09-20 11:11:11.9999');
```
Comment (SeaRise) [inline tests/fullstack-test/expr/cast_as_duration.test:22]: And add null value?

File: dbms/src/Functions/tests/gtest_tidb_conversion.cpp [hunk 5]
```diff
@@ -1749,5 +1749,63 @@ try
...
+                                                     (11 * 3600 + 11 * 60 + 11) * 1000000000L + 110000000L,
+                                                     (11 * 3600 + 11 * 60 + 11) * 1000000000L + 120000000L,
+                                                     (11 * 3600 + 11 * 60 + 12) * 1000000000L + 000000000L})
+            .column,
+        to_type_3,
+        "datetime_output3");
+
+
+    // ASSERT_COLUMN_EQ(datetime_output1, executeFunction(func_name, {ctn_datetime, createCastTypeConstColumn(to_type_1->getName())}));
+    // ASSERT_COLUMN_EQ(datetime_output2, executeFunction(func_name, {ctn_datetime, createCastTypeConstColumn(to_type_2->getName())}));
```
Comment (SeaRise) [inline dbms/src/Functions/tests/gtest_tidb_conversion.cpp:1805]: Why not execute it?

File: dbms/src/Functions/tests/gtest_tidb_conversion.cpp [hunk 6]
```diff
@@ -1749,5 +1749,63 @@ try
 }
 CATCH
 
+TEST_F(TestTidbConversion, castTimeAsDuration)
```
Comment (SeaRise) [inline dbms/src/Functions/tests/gtest_tidb_conversion.cpp:1752]: and add tests for const and null_value?

File: dbms/src/Functions/FunctionsTiDBConversion.h (hunk 3 above)
Comment (AntiTopQuark) [reply dbms/src/Functions/FunctionsTiDBConversion.h]: thx，done

File: dbms/src/Functions/tests/gtest_tidb_conversion.cpp (hunk 5 above)
Comment (AntiTopQuark) [reply dbms/src/Functions/tests/gtest_tidb_conversion.cpp]: sorry，done

File: tests/fullstack-test/expr/cast_as_duration.test (hunk 4 above)
Comment (AntiTopQuark) [reply tests/fullstack-test/expr/cast_as_duration.test:22]: dong

File: dbms/src/Functions/tests/gtest_tidb_conversion.cpp (hunk 6 above)
Comment (AntiTopQuark) [reply dbms/src/Functions/tests/gtest_tidb_conversion.cpp:1752]: done

**Review by SeaRise — COMMENTED** [review]
basically lgtm

Add release note like 'Support to pushdown CastTimeAsDuration to TiFlash'.
And add the pr link of tidb?

**Comment by AntiTopQuark** [comment]
the pr link of tidb:
https://github.com/pingcap/tidb/pull/37036

**Comment by SeaRise** [comment]
> the pr link of tidb: [pingcap/tidb#37036](https://github.com/pingcap/tidb/pull/37036)

like https://github.com/pingcap/tiflash/pull/5156
twMbPCDlxd
OLhideZvOo

**Comment by AntiTopQuark** [comment]
> > the pr link of tidb: [pingcap/tidb#37036](https://github.com/pingcap/tidb/pull/37036)
> 
> like #5156 twMbPCDlxd OLhideZvOo

done,thx

**Comment by SeaRise** [comment]
> > > the pr link of tidb: [pingcap/tidb#37036](https://github.com/pingcap/tidb/pull/37036)
> > 
> > 
> > like #5156 twMbPCDlxd OLhideZvOo
> 
> done,thx

Also please add release notes. like
twMbPCDlxd

**Comment by AntiTopQuark** [comment]
> > > > the pr link of tidb: [pingcap/tidb#37036](https://github.com/pingcap/tidb/pull/37036)
> > > 
> > > 
> > > like #5156 twMbPCDlxd OLhideZvOo
> > 
> > 
> > done,thx
> 
> Also please add the release note. like twMbPCDlxd

done

**Review by SeaRise — APPROVED** [review]
LGTM

File: dbms/src/Functions/FunctionsTiDBConversion.h [hunk 7]
```diff
@@ -1610,6 +1611,47 @@ struct TiDBConvertToDuration
...
+        {
+            // cast date as duration
+            auto to_col = ColumnVector<Int64>::create();
+            auto & vec_to = to_col->getData();
+            vec_to.resize(size);
+            for (size_t i = 0; i < size; ++i)
+            {
+                vec_to[i] = 0; // The DATE type is used for values with a date part but no time part. The value of Duration is always zero.
+            }
+            block.getByPosition(result).column = std::move(to_col);
```
Comment (Willendless) [inline dbms/src/Functions/FunctionsTiDBConversion.h:1624]: Since the result column is always 0, can we use ColumnConst to optimize a little bit?

Comment (AntiTopQuark) [reply dbms/src/Functions/FunctionsTiDBConversion.h:1624]: thx,done

Comment (AntiTopQuark) [reply dbms/src/Functions/FunctionsTiDBConversion.h:1624]: thx,done

Comment (Willendless) [reply dbms/src/Functions/FunctionsTiDBConversion.h:1615]: `block.getByPosition(result).column = std::make_shared<DataTypeInt64>()->createColumnConst(size, 0);`

File: dbms/src/Functions/FunctionsTiDBConversion.h [hunk 8]
```diff
@@ -1601,6 +1602,47 @@ struct TiDBConvertToDuration
...
```
Comment (Willendless) [inline dbms/src/Functions/FunctionsTiDBConversion.h:1615]: Maybe use `block.getByPosition(result).column = DataTypeInt64()->createColumnConst(size, 0);` instead?

Comment (AntiTopQuark) [reply dbms/src/Functions/FunctionsTiDBConversion.h]: done，thx~

**Review by ywqzzy — APPROVED** [review]
LGTM

**Pull request merged**